contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -367,6 +541,173 @@ process_remove_all(struct sshbuf* request, struct sshbuf* response, struct agent
 	return r;
 }
 
+#ifdef ENABLE_PKCS11
+int process_add_smartcard_key(struct sshbuf* request, struct sshbuf* response, struct agent_connection* con)
+{
+	char *provider = NULL, *pin = NULL, canonical_provider[PATH_MAX];
+	int i, count = 0, r = 0, request_invalid = 0, success = 0;
+	struct sshkey **keys = NULL;
+	struct sshkey* key = NULL;
+	size_t pubkey_blob_len, provider_len, pin_len, epin_len;
+	u_char *pubkey_blob = NULL;
+	char *thumbprint = NULL;
+	char *epin = NULL;
+	HKEY reg = 0, sub = 0, user_root = 0;
+	SECURITY_ATTRIBUTES sa = { 0, NULL, 0 };
+
+	if ((r = sshbuf_get_cstring(request, &provider, &provider_len)) != 0 ||
+		(r = sshbuf_get_cstring(request, &pin, &pin_len)) != 0) {
+		debug("add smartcard request is invalid");
+		request_invalid = 1;
+		goto done;
+	}
+
+	if (realpath(provider, canonical_provider) == NULL) {
+		debug("failed PKCS#11 add of \"%.100s\": realpath: %s",
+			provider, strerror(errno));
+		request_invalid = 1;
+		goto done;
+	}
+
+	// Remove 'drive root' if exists
+	if (canonical_provider[0] == '/')
+		memmove(canonical_provider, canonical_provider + 1, strlen(canonical_provider));
```

Comment:
Just a side note: wouldn't it be easier to have a (buffer, pointer) pair and just increment the pointer if needed?

Reply:
Easier as in faster/more efficient?
I think that it's pretty readable that way, but I'm open to changing that if you have a strong preference.